src/Functions/FunctionsJSON.h

```diff
@@ -286,27 +288,81 @@ class FunctionJSON : public IFunction, WithContext
     String getName() const override { return Name::name; }
     bool isVariadic() const override { return true; }
     size_t getNumberOfArguments() const override { return 0; }
+
+    // Both NULL and JSON NULL should generate NULL value.
+    bool useDefaultImplementationForNulls() const override { return false; }
+
     bool useDefaultImplementationForConstants() const override { return true; }
     bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
 
     DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
     {
+        NullPresence null_presence = getNullPresense(arguments);
+
+        if (null_presence.has_null_constant)
+        {
+            return makeNullable(std::make_shared<DataTypeNothing>());
+        }
+        if (null_presence.has_nullable)
+        {
+            auto nested_columns = createBlockWithNestedColumns(arguments);
+            auto return_type = Impl<DummyJSONParser>::getReturnType(Name::name, nested_columns);
+            return makeNullable(return_type);
+        }
         return Impl<DummyJSONParser>::getReturnType(Name::name, arguments);
     }
 
     ColumnPtr executeImpl(const ColumnsWithTypeAndName & arguments, const DataTypePtr & result_type, size_t input_rows_count) const override
     {
+        NullPresence null_presence = getNullPresense(arguments);
```

Comment:
@amosbird Can we separate the code of this function into two parts - one to handle nulls, and one to call `Executor::run()`? And the same for `getReturnTypeImpl()`? Just to make the code clearer.

Reply:
Sure. In fact let's build a OverloadResolver and store `NullPresence` beforehand.